pkg/listwatch/listwatch.go

```diff
@@ -0,0 +1,194 @@
+// If the slice of namespaces contains only v1.NamespaceAll, then this func assumes that the client
+// has List and Watch privileges and returns a regular cache.ListWatch, since there is no other way to get all namespaces.
+func NewFilteredUnprivilegedNamespaceListWatchFromClient(c cache.Getter, namespaces []string, optionsModifier func(options *metav1.ListOptions)) *cache.ListWatch {
+	// If the only namespace given is `v1.NamespaceAll`, then this cache.ListWatch must be privileged.
+	// In this case, return a regular cache.ListWatch.
+	if IsAllNamespaces(namespaces) {
+		return cache.NewFilteredListWatchFromClient(c, "namespaces", metav1.NamespaceAll, optionsModifier)
+	}
+	listFunc := func(options metav1.ListOptions) (runtime.Object, error) {
+		optionsModifier(&options)
+		list := &v1.NamespaceList{}
+		for _, name := range namespaces {
+			result := &v1.Namespace{}
+			err := c.Get().
+				Resource("namespaces").
+				Name(name).
+				VersionedParams(&options, scheme.ParameterCodec).
+				Do().
+				Into(result)
+			if err != nil {
+				return nil, err
+			}
+			list.Items = append(list.Items, *result)
+		}
+		return list, nil
+	}
+	watchFunc := func(_ metav1.ListOptions) (watch.Interface, error) {
+		return watch.NewFake(), nil
+	}
+	return &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// MultiNamespaceListerWatcher takes a list of namespaces and a cache.ListerWatcher generator func
+// and returns a single cache.ListerWatcher capable of operating on multiple namespaces.
+func MultiNamespaceListerWatcher(namespaces []string, f func(string) cache.ListerWatcher) cache.ListerWatcher {
+	// If there is only one namespace then there is no need to create a proxy.
+	if len(namespaces) == 1 {
+		return f(namespaces[0])
+	}
+	var lws []cache.ListerWatcher
+	for _, n := range namespaces {
+		lws = append(lws, f(n))
+	}
+	return multiListerWatcher(lws)
+}
+
+// multiListerWatcher abstracts several cache.ListerWatchers, allowing them
+// to be treated as a single cache.ListerWatcher.
+type multiListerWatcher []cache.ListerWatcher
+
+// List implements the ListerWatcher interface.
+// It combines the output of the List method of every ListerWatcher into
+// a single result.
+func (mlw multiListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	l := metav1.List{}
+	var resourceVersions []string
+	for _, lw := range mlw {
+		list, err := lw.List(options)
+		if err != nil {
+			return nil, err
+		}
+		items, err := meta.ExtractList(list)
+		if err != nil {
+			return nil, err
+		}
+		metaObj, err := meta.ListAccessor(list)
+		if err != nil {
+			return nil, err
+		}
+		for _, item := range items {
+			l.Items = append(l.Items, runtime.RawExtension{Object: item.DeepCopyObject()})
+		}
+		resourceVersions = append(resourceVersions, metaObj.GetResourceVersion())
+	}
+	l.ListMeta.ResourceVersion = strings.Join(resourceVersions, "/")
```

Comment:
let's put a comment on here why we're doing this

Reply:
:+1: